File: Core/src/main/java/jokatu/game/input/SingleKeyInputDeserialiser.java
package jokatu.game.input;

import org.jetbrains.annotations.NotNull;

import java.util.Map;

/**
 * An {@link InputDeserialiser} where the JSON is expected to have only one key.
 * @param <I> the type of the {@link Input} to output
 */
public abstract class SingleKeyInputDeserialiser<I extends Input> extends InputDeserialiser<I> {
	@NotNull
	@Override
	public final I deserialise(@NotNull Map<String, Object> json) throws DeserialisationException {
		String keyName = getKeyName();
		getMandatoryKeyValue(json, keyName);
		Object value = json.get(keyName);
		return deserialiseSingleValue(json, value);
	}

	@NotNull
	protected abstract I deserialiseSingleValue(@NotNull Map<String, Object> json, @NotNull Object value) throws DeserialisationException;

	@NotNull
	protected abstract String getKeyName();
}

File: Core/src/main/kotlin/jokatu/game/input/SingleKeyInputDeserialiser.kt
package jokatu.game.input

/**
 * An [InputDeserialiser] where the JSON is expected to have only one key.
 * @param <I> the type of the [Input] to output
 */
abstract class SingleKeyInputDeserialiser<I : Input>(protected val keyName: String) : InputDeserialiser<I>() {
	@Throws(DeserialisationException::class)
	override fun deserialise(json: Map<String, Any>): I {
		if (json.size > 1) {
			throw DeserialisationException(json, "There was more than one key.")
		}
		val value = getMandatoryKeyValue(keyName, json)
		return deserialiseSingleValue(json, value)
	}

	@Throws(DeserialisationException::class)
	protected abstract fun deserialiseSingleValue(json: Map<String, Any>, value: Any): I
}
